pygerm/reduction.py

```diff
@@ -1,144 +1,297 @@
+                   calibration = None,
+                   td_resolution=40e-9, n_chans=32, n_chips=12,
+                   jump_bits=29, thresh_bits=26,
+                   chunksize=1000000,
+                   plot=True, verbose=False):
+    '''
+        Histogram the GeRM data.
+
+        Parameters
+        ----------
+
+        germ_ts : dask.array or np.ndarray
+            timestamp in seconds
+        germ_td : dask.array or np.ndarray
+            timestamp in clock cycles
+        germ_pd : dask.array or np.ndarray
+            the germ data
+        germ_chip : dask.array or np.ndarray
+            the channel number
+        germ_chan : dask.array or np.ndarray
+            the chip number
+
+        time_resolution: the time to bin by
+        start_time : the start time
+        end_time : the stop time
+
+        energy_resolution: int
+            the number of energies to count on
+        min_energy: float
+            the minimum energy to measure up to
+        max_energy: float
+            the maximum energy to measure up to
+
+        td_resolution: float
+            the time resolution of each count in germ_td in seconds
+        n_chans: int
+            the number of channels
+        n_chips : int
+            the number of chips
+
+        calibration: 2d np.ndarray
+            calibration matrix for the data
+            It should be a 2d array of dimensions 2 x (n_chips*n_chans + 1)
+            It specifies an interpolation of the form:
+                res = res*cal[0, loc] + cal[1, loc]
+
+        jump_bits: int
+            Don't change this unless you know what you're doing
+            (see fixtime)
+
+        thresh_bits: int
+            Don't change this unless you know what you're doing
+            (see fixtime)
+
+        chunksize: int
+            The number of chunks of data to read.
+            Note: specify this for large data sets that can't fit into memory!
+            Keep increasing until it fits. The smaller the chunksize, the
+            slower the analysis (more iterations required).
+
+        verbose: bool, optional
+            print more verbose flags if needed
+
+        Notes
+        -----
+            Either *all* of [germ_ts, germ_td, germ_pd, germ_chip, germ_chan]
+            are dask arrays or *None* are dask arrays.
+
+        Returns
+        -------
+        spectrum : 3d np.ndarray
+            This is a histogram of by chip x energy x time
+
+        spectrum_edges: list of 3 1d np.ndarrays
+            List of three edges, in order:
+                - the chips (useful for verification that the edges are correct,
+                should be integers)
+                - the energies (should be in units of keV)
+                    note: a proper calibration matrix must be used to properly
+                    convert from ADU to keV. Else, the result is not keV
+                - the times (in s)
+
+    '''
+    if hasattr(germ_ts, 'compute'):
```

Comment:
There has to be a way to do this with out resorting to `hasattr` checks?  I think `np.asarray` will be lazy for numpy arrays and force comptuation of dask arrays (or anything else that respects the `__array__` protocol).

Reply:
this is perfect and simple, tested and worked thanks!